File: lib/emulation_system/emulation/bot/state.rb
module EmulationSystem
  module Emulation
    class Bot
      # === Состояние бота
      # * +pos+ - позиция бота на поле
      # * +angle+ - направление движения бота
      # * +speed+ - модуль мгновенной скорости
      # * +desired_speed+ - конечное значение модуля скорости при разгоне и торможении
      # * +health+ - здоровье
      class State < Struct.new :pos, :angle, :speed, :desired_speed, :health
        def radians
          angle * Math::PI / 180
        end
        
        def radians=(value)
          self.angle = value / Math::PI * 180
        end

        def cosa
          Math::cos(radians)
        end

        def sina
          Math::sin(radians)
        end

        # Возвращает режим движения:
        # * +accelerated+ - разгоняется
        # * +decelerated+ - тормозит
        # * +uniform+     - равномерное движение
        def speed_mode
          if speed > desired_speed
            :decelerated
          elsif speed < desired_speed
            :accelerated
          else
            :uniform
          end
        end

        # Корректирует значения координат, скорости и здоровья
        # если они выходят за пределы
        def correct
          self.pos.x  = correct_value pos.x,  World::BOT_RADIUS, World::FIELD_SIZE - World::BOT_RADIUS
          self.pos.y  = correct_value pos.y,  World::BOT_RADIUS, World::FIELD_SIZE - World::BOT_RADIUS
          self.speed  = correct_value speed,  0, World::MAX_SPEED
          self.health = correct_value health, 0, World::MAX_HEALTH
        end

        # Возвращает значение +val+, ограниченное
        # до пределов +min+..+max+
        def correct_value(val, min, max)
          val < min ? min : (val > max ? max : val)
        end
      end
    end
  end
end

File: ext/state.c
#include "state.h"
#include <math.h>
#include <ruby.h>

double _radians2degrees(double angle)
{
    return angle/M_PI*180;
}

double _degrees2radians(double angle)
{
    return angle*M_PI/180;
}
#define CORRECT_VALUE(val, min, max) val = (val < min) ? min : (val > max ? max : val)

State * state_new()
{
    State * s = (State *)xmalloc(sizeof(State));
    s->health = WORLD_MAX_HEALTH;
    s->speed = s->desired_speed = 0;
    s->pos = v2d_new();
    return s;
}

void state_delete(State * s)
{
    v2d_delete(s->pos);
    xfree(s);
}

double state_cosa(State * s)
{
    return cos( s->radians );
}

double state_sina(State * s)
{
    return sin( s->radians );
}

void state_correct(State * s)
{
    CORRECT_VALUE(s->pos->x, WORLD_BOT_RADIUS, WORLD_FIELD_SIZE - WORLD_BOT_RADIUS);
    CORRECT_VALUE(s->pos->y, WORLD_BOT_RADIUS, WORLD_FIELD_SIZE - WORLD_BOT_RADIUS);
    CORRECT_VALUE(s->speed, 0, WORLD_MAX_SPEED);
    CORRECT_VALUE(s->health, 0, WORLD_MAX_HEALTH);
}

SpeedMode state_speed_mode(State *s)
{
    if(s->speed > s->desired_speed)
        return SM_DECELERATED;
    else if(s->speed < s->desired_speed)
        return SM_ACCELERATED;
    else
        return SM_UNIFORM;
}

